### ISSUE TITLE
Locally declared @Inherited annotations are not present in native image

### ISSUE THREAD

**Issue by sbrannen** [oracle/graal#2243]
# Overview

Given an `@Inherited` annotation that is declared on a superclass and again on a subclass, the annotation is not _present_ as a "locally declared" annotation on the subclass within a native image.

This breaks code that relies on the actual location of the declaration of the annotation.

For example, given the `InheritedAnnotations` test case below, the output when using a standard JDK results in the following.

```diff
+ SUCCESS: @InheritedAnnotation was found as an annotation on Superclass.
+ SUCCESS: @InheritedAnnotation was found as a declared annotation on Superclass.
+ SUCCESS: @InheritedAnnotation was found as an annotation on Subclass.
+ SUCCESS: @InheritedAnnotation was found as a declared annotation on Subclass.
```

Whereas, if you compile the test case into a native image, the output is the following.

```diff
+ SUCCESS: @InheritedAnnotation was found as an annotation on Superclass.
+ SUCCESS: @InheritedAnnotation was found as a declared annotation on Superclass.
+ SUCCESS: @InheritedAnnotation was found as an annotation on Subclass.
- FAILURE: @InheritedAnnotation was NOT found as a declared annotation on Subclass.
```

The above **FAILURE** points out the bug within a native image.

See https://github.com/sbrannen/graalvm-playground/tree/master/inherited_annotations for full details.

```java
import java.lang.annotation.Annotation;
import java.lang.annotation.Inherited;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

public class InheritedAnnotations {

	public static void main(String[] args) {
		findAnnotation(Superclass.class);
		findAnnotation(Subclass.class);
	}

	private static void findAnnotation(Class<?> clazz) {
		Annotation[] annotations = clazz.getAnnotations();
		if (annotations.length == 1 && annotations[0].annotationType().equals(InheritedAnnotation.class)) {
			System.out.format("SUCCESS: @InheritedAnnotation was found as an annotation on %s.%n",
				clazz.getSimpleName());
		}
		else {
			System.out.format("FAILURE: @InheritedAnnotation was NOT found as an annotation on %s.%n",
				clazz.getSimpleName());
		}
		Annotation[] declaredAnnotations = clazz.getDeclaredAnnotations();
		if (declaredAnnotations.length == 1
				&& declaredAnnotations[0].annotationType().equals(InheritedAnnotation.class)) {
			System.out.format("SUCCESS: @InheritedAnnotation was found as a declared annotation on %s.%n",
				clazz.getSimpleName());
		}
		else {
			System.out.format("FAILURE: @InheritedAnnotation was NOT found as a declared annotation on %s.%n",
				clazz.getSimpleName());
		}
	}
}
```

```java
@Retention(RetentionPolicy.RUNTIME)
@Inherited
@interface InheritedAnnotation {
}
```

```java
@InheritedAnnotation
class Superclass {
}
```

```java
@InheritedAnnotation
class Subclass extends Superclass {
}
```

# Environment

- GraalVM version: CE 20.0
- JDK major version: 8
- OS: macOS 10.14.6
- Architecture: i386 / x86_64

**Comment by eginez** [comment]
There seems to be a bug in the `getDeclaredAnnotations` API. Conversly both `getAnnotations()` and `getAnnotations(Class<?>)` does report the expected result

**Comment by eginez** [comment]
should be fixed with https://github.com/oracle/graal/commit/c78d5fd611f8bfc1f449f7f7655f44644497fb55

**Comment by sbrannen** [comment]
Great news, @eginez!

I'll try it out...

**Comment by sbrannen** [comment]
I can confirm that the example now works with GraalVM CE 20.1.0-dev-20200411_0336.

**Comment by sbrannen** [comment]
Well, I spoke too soon.

The change in the commit did in fact cause the original example in this issue to pass; however, it causes 33 failures in the native image test run for `spring-core` in the Spring Framework.

If you change the examples to look like this (where the `@Inherited` annotation is only present on the superclass)...

```java
@Retention(RetentionPolicy.RUNTIME)
@Inherited
@interface InheritedAnnotation {
}

@InheritedAnnotation
static class Superclass {
}

static class Subclass extends Superclass {
}
```

Then the output in a normal JDK is:

```diff
+ @InheritedAnnotation was found as an annotation on Superclass.
+ @InheritedAnnotation was found as a declared annotation on Superclass.
+ @InheritedAnnotation was found as an annotation on Subclass.
+ @InheritedAnnotation was NOT found as a declared annotation on Subclass.
```

But the output within a native image is now:

```diff
+ @InheritedAnnotation was found as an annotation on Superclass.
+ @InheritedAnnotation was found as a declared annotation on Superclass.
+ @InheritedAnnotation was found as an annotation on Subclass.
- @InheritedAnnotation was found as a declared annotation on Subclass.
```

From the above, we can see that an `@Inherited` annotation is now considered to be "directly present" on the subclass even if it's only _inherited_.

Thus, please reopen this issue.

**Comment by eginez** [comment]
Ok a new commit to address this issue has been merged: 83ac94f5b0a67044cbf3a59569861a0ee93d85f2

**Comment by sbrannen** [comment]
Wow, that commit is considerably more involved!

Sorry it ended up being so much work, though by the looks of it... I suppose the necessary information simply wasn't previously present in the native image.

> Ok a new commit to address this issue has been merged: 83ac94f

When will I be able to try that out with one of the downloads from [here](https://github.com/graalvm/graalvm-ce-dev-builds/releases/)?

**Comment by eginez** [comment]
The latest dev build now contains this fix. eg: https://github.com/graalvm/graalvm-ce-dev-builds/releases/download/20.1.0-dev-20200419_0207/graalvm-ce-java8-darwin-amd64-20.1.0-dev.tar.gz
